### ISSUE TITLE
How to exclude paths for features such as jump-to-definition?

### ISSUE THREAD

**Issue by jordan-yee** [kak-lsp/kak-lsp#626]
When attempting to use features such as jump-to-definition or find-references, I'm getting results from the project's output folder rather than the actual source code. Is there a way to get kak-lsp to ignore such directories?

I'm facing this problem in a ClojureScript project, but I imagine the same problem would exist in other languages, particularly web languages.

A related issue for that particular language server: https://github.com/clojure-lsp/clojure-lsp/issues/185

**Comment by jordan-yee** [comment]
Thanks for quick response!

I too would have expected an ignore-output-directory setting to exist on the server side, as that is what deals with the language specifics. There is actually a [setting](https://clojure-lsp.io/settings/#all-settings) for the clojure-lsp server that sounds like it should be doing the right thing by default. But since that isn't what I'm experiencing in practice, and since someone in [that linked issue](https://github.com/clojure-lsp/clojure-lsp/issues/185) stated that they thought this should be up to the client--here I am.

It sounds like the short answer here is that kak-lsp does not support this and has no plans to. That makes sense as far as my limited understanding of the inner workings of LSP goes, but I had to at least check.

**Comment by krobelus** [comment]
> There is actually a [setting](https://clojure-lsp.io/settings/#all-settings) for the clojure-lsp server that sounds like it should be doing the right thing by default.

which setting?

> kak-lsp does not support this and has no plans to

if we knew how to do, we would probably support it

Can you give minimal steps to reproduce the problem? (Probably trivial but I've never used clojure).
I can try if VSCode has the same problem.
If not, it must be something they're doing on the client side, and we can probably mimic that.
Otherwise, we should probably complain to clojure-lsp to fix this for everyone.

**Comment by krobelus** [comment]
> I'm looking into solutions on the lsp server end. In this [Slack thread](https://clojurians.slack.com/archives/CPABC1H61/p1657820971604899), they are requesting `server info`, which other clients provide commands for. Is there a command for this via kak-lsp? I don't see one for it in the editor or readme.
>
>		> we need the server info information, emacs and calva have commands to retrieve that

That link requires sign up and it doesn't tell me how to create an account.
"server info" doens't ring a bell but I guess they mean initializationOptions.

Anyway, I managed to reproduce and fix a similar problem.

TL;DR: you can set `:source-paths-ignore-regex` in one of 4 different ways, see https://clojure-lsp.io/settings/

Here's what I did

0. Install clojure, leiningen and clojure-lsp
1. `lein new figwheel hello-world`
2. cd `hello-world`
3. Define a function in another namespace

	$ mkdir src/other && cat > src/other/other.cljs << EOF
	(ns hello-world.other)

	(defn myspecialfunction [])
	EOF

4. Call the function from `src/hello_world/core.cljs`

	(defn foo []
		(myspecialfunction))

  Observe that `myspecialfunction` has an error (marked with the red * in the gutter by kak-lsp), because it's not in scope.
  This is easy to fix because clojure-lsp provides the appropriate code action.
  Use `enter-user-mode lsp` press `a` and select `Add require suggestion`. Now the top paragraph should look like.

	(ns hello-world.core
	    (:require
	     [hello-world.other :refer [myspecialfunction]]))

5. Add a second source directory that sorts before `src` and also provides `myspecialfunction`. This simulates your output directory.

	$ mkdir asrc
	$ cp src/other asrc/other

  To have the build system pick it up, change the line `:source-paths ["src"]` in `project.clj` to `:source-paths ["src", "asrc"]`.

6. run `lein figwheel` to build everything (probably not necessary though)

7. Now we can reproduce the problem: run `kak src/hello_world/core.cljs` and run `gd` on `myspecialfunction`. This will jump to `asrc/output/other.cljs` (but we want `src/output/other.cljs`).

8. Since [clojure-lsp has added `:source-paths-ignore-regex`](https://github.com/clojure-lsp/clojure-lsp/releases/tag/2022.06.22-14.09.50) we can use that to avoid `asrc`.

	$ cat >? .lsp/config.edn << EOF
	{:source-paths-ignore-regex
		["resources.*" "target.*" ;; defaults
		"asrc.*" ;; added
		]
	}
	EOF

9. After restarting Kakoune (or clojure-lsp), we successfully jump to `src/output/other.cljs`

This fix works independently of Kakoune/kak-lsp, so it should help VSCode users too.

FWIW, if you want to use [LSP settings](https://clojure-lsp.io/settings/#initialization-options) instead of ` .lsp/config.edn` you can do that too.
You just need to add 3 lines to your `kak-lsp.toml`.

```toml
[language.clojure]
filetypes = ["clojure"]
roots = ["project.clj"]
command = "clojure-lsp"
settings_section = "_"
[language.clojure.settings._]
source-paths-ignore-regex = ["resources.*", "target.*", "asrc.*"]
```

---

I suspect that you are seeing this issue due to a non-standard project setup.
IIUC by default, output artifacts are placed in `resources` or `target`, which are ignored by default.

**Comment by jordan-yee** [comment]
Wow, nice work! So it would seem source directories are scanned alphabetically--good to note.

> 5. Add a second source directory that sorts before src and also provides myspecialfunction. This simulates your output directory.

One important detail I would add to this step is that the second directory *should* have to include a definition for the same namespace with the same function name, i.e. copy the first file you defined into the second directory. If the lsp server would pick up a function with the same name in a different namespace, then we'd have much bigger problems...

> Observe that myspecialfunction has an error (marked with the red * in the gutter by kak-lsp), because it's not in scope.
> This is easy to fix because clojure-lsp provides the appropriate code action.
> Use enter-user-mode lsp press a and select Add require suggestion.

Yes! Great feature that is.

*Out of scope for this issue, but while it's fresh I'd like to mention an issue with that feature. I used to see a list of possible namespace suggestions to require, but recently it seems to have been just using the first one. Maybe this has to do with changes to lsp-code-actions.*

Now, back to the issue at hand...

So I figured out that I was using a version of clojure-lsp that didn't yet have the `:source-paths-ignore-regex` setting. I thought that the issue was persisting after updating this, but that was only partially the case. It did seem to fix the go-to-definition command, BUT I still see files from the output directory in the results of the **list-symbol-references** command. This one feels like it could be a bug, unless that feature cues off a different setting.

**Comment by krobelus** [comment]
> I used to see a list of possible namespace suggestions to require, but recently it seems to have been just using the first one. Maybe this has to do with changes to lsp-code-actions.*

Unlikely. We just forward code actions from the server. Maybe the server behavior changed? LMK if you have a reproducer.

> BUT I still see files from the output directory in the results of the **list-symbol-references** command. This one feels like it could be a bug, unless that feature cues off a different setting.

I can't reproduce, `lsp-references` does respect ignored source paths here
